rustls-cert-gen/src/args.rs

```diff
@@ -0,0 +1,95 @@
+//! Command Line argument parsing
+#![allow(missing_docs)]
+
+use std::{net::IpAddr, path::PathBuf};
+
+use bpaf::Bpaf;
+use rcgen::SanType;
+
+#[derive(Debug, Clone, Bpaf)]
+#[bpaf(fallback(Purpose::Omit))]
+/// Extended Key Usage Purpose (rfc5280, section 4.2.1.12)
+pub enum Purpose {
+	/// Extended Key Usage Purpose: ClientAuth
+	ClientAuth,
+	/// Extended Key Usage Purpose: ServerAuth
+	ServerAuth,
+	#[bpaf(skip)]
+	/// Default case is to omit adding Extended Key Usage Purpose
+	Omit,
+}
+
+#[derive(Clone, Debug, Bpaf)]
+#[bpaf(options)]
+/// rustls-cert-gen Tls Certificate Generator
+pub struct Options {
+	/// Output Directory for generated files
+	#[bpaf(short, long, argument("output/path/"))]
+	pub output: PathBuf,
+	/// Persist ca signing-key to `output` dir
+	#[bpaf(long, switch)]
+	pub insecure: bool,
+	/// Signature/Encryption algorithm
+	#[bpaf(short, long, fallback("pkcs_ecdsa_p256_sha256".into()), display_fallback)]
+	pub sig_algo: String,
+	#[bpaf(external)]
+	/// Extended Key Usage Purpose
+	pub purpose: Purpose,
+	/// Basename for end-entity cert/key
+	#[bpaf(long, fallback("cert".into()), display_fallback)]
+	pub cert_file_name: String,
+	/// Basename for ca cert/key
+	#[bpaf(long, fallback("root-ca".into()), display_fallback)]
+	pub ca_file_name: String,
+	/// Subject Alt Name (apply multiple times for multiple names/Ips)
+	#[bpaf(many, long, argument::<String>("san"), map(parse_sans))]
+	pub san: Vec<SanType>,
+	/// Common Name (Currently only used for end-entity)
+	#[bpaf(long, fallback("Tls End-Entity Certificate".into()), display_fallback)]
+	pub common_name: String,
+	/// Country Name (Currently only used for ca)
+	#[bpaf(long, fallback("BR".into()), display_fallback)]
+	pub country_name: String,
+	/// Organization Name (Currently only used for ca)
+	#[bpaf(long, fallback("Crab widgits SE".into()), display_fallback)]
+	pub organization_name: String,
+}
+
+/// Parse cli input into SanType. Try first `IpAddr`, if that fails
+/// declare it to be a DnsName.
+fn parse_sans(hosts: Vec<String>) -> Vec<SanType> {
+	hosts
+		.into_iter()
+		.map(|s| s.to_string())
```

Comment:
This isn't needed if you make `parse_san` take a `&str` instead of `String`.

Reply:
In that case I think we would need something like `.iter().map(|s|parse_san(s))`. The function call is needed to get the `&str` (I think). But it turns out that `s.to_string()` was unnecessary even without changing the signature. So I just removed that line in the latest commit.